### PR TITLE
feat(Channel): improve typeguards

### PR DESCRIPTION
**Pull request by ImRodry**
**Please describe the changes this PR makes and why it should be merged:**
This PR creates typeguards for every channel type and renames the existing ones to isTextBased and isVoiceBased, as this was already what they did, allowing for accurate isText and isVoice typeguards.

**Status and versioning classification:**
- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating
- This PR changes the library's interface (methods or parameters added)
- This PR includes breaking changes (methods removed or renamed, parameters moved or removed)

### CODE REVIEW

File: src/structures/Channel.js
Comment (papsavas) [inline src/structures/Channel.js:171]: Wouldn't `this.type.endsWith('_THREAD')` or atleast `includes('THREAD')` be more suitable according to their [naming](https://github.com/discord/discord-api-docs/blob/master/docs/resources/Channel.md?plain=1#L54-L56)?

Comment (ImRodry) [reply src/structures/Channel.js:171]: What if they add more thread types that don’t end with _THREAD? Unlikely, but still possible and I don’t think a thread would ever not have autoArchiveDuration, however I’m happy to revert if the maintainers think I should, just did that for consistency with the isTextBased method

Comment (papsavas) [reply src/structures/Channel.js:171]: A defined `parentId` can also indicate a channel's category. Perhaps a combination of checks is safer

Comment (ImRodry) [reply src/structures/Channel.js:171]: I can always revert to the thread channel types check, but I think it’d be best to use the same check across all methods

Comment (vladfrangu) [reply src/structures/Channel.js:171]: I agree with @papsavas here, checking `autoArchiveDuration` makes this type guard feel out of place... Either use an array of thread types and check if this.type is present (ideal), or check if the type includes THREAD (less ideal but still better)

Comment (ImRodry) [reply src/structures/Channel.js:171]: Alright I’ll change these ones to check for the type and keep the isXBased ones checking for a property in the channel

File: src/structures/Channel.js
Comment (papsavas) [inline src/structures/Channel.js:115]:
```suggestion
    return this.type === ChannelTypes.GUILD_TEXT;
```

File: src/structures/Channel.js
Comment (papsavas) [inline src/structures/Channel.js:123]:
```suggestion
    return this.type === ChannelTypes.DM;
```

File: src/structures/Channel.js
Comment (papsavas) [inline src/structures/Channel.js:131]:
```suggestion
    return this.type === ChannelTypes.GUILD_VOICE;
```

File: src/structures/Channel.js
Comment (papsavas) [inline src/structures/Channel.js:139]:
```suggestion
    return this.type === ChannelTypes.GROUP_DM;
```

File: src/structures/Channel.js
Comment (papsavas) [inline src/structures/Channel.js:147]:
```suggestion
    return this.type === ChannelTypes.GUILD_CATEGORY;
```

File: src/structures/Channel.js
Comment (papsavas) [inline src/structures/Channel.js:155]:
```suggestion
    return this.type === ChannelTypes.GUILD_NEWS;
```

File: src/structures/Channel.js
Comment (papsavas) [inline src/structures/Channel.js:163]:
```suggestion
    return this.type === ChannelTypes.GUILD_STORE;
```

File: src/structures/Channel.js
Comment (papsavas) [inline src/structures/Channel.js:179]:
```suggestion
    return this.type === ChannelTypes.GUILD_STAGE_VOICE;
```

**Review by papsavas — CHANGES_REQUESTED** [review]
Feels like this would be preferred rather than raw strings

File: src/structures/Channel.js
Comment (vladfrangu) [reply src/structures/Channel.js:115]: I'm all for enum usages but this would never return true... pretty sure that ChannelTypes.GUILD_TEXT is a number, not a string

Comment (papsavas) [reply src/structures/Channel.js:115]: I tested this and you're right. I will change it to ```ChannelTypes[ChannelTypes.X]``` for now but perhaps a function for this would be a little more subtle

Comment (suneettipirneni) [reply src/structures/Channel.js:115]: I mean, to be fair using an indexed access here seems like a lot more verbose, and it (maybe...?) has performance implications rather than using a string comparison.

Comment (papsavas) [reply src/structures/Channel.js:115]: All (afaik) true, but i still believe that raw strings are not the way to go. If discord decides to rename channel types for some reason (especially since they work with int numbers so all good), it gets messy. We should try and minimize the exposure of raw strings all around the project and gather them in the least possible files needed. This way, something changes, you change it once.

Comment (ImRodry) [reply src/structures/Channel.js:115]: If the type name changed you’d be having to change the name of the method most likely anyway. Isn’t this solution much worse performance wise? And I mean you still have the channel type string there so it doesn’t improve anything in that regard

**Review by vladfrangu — APPROVED** [review]

**Review by iCrawl — APPROVED** [review]

**Review by SpaceEEC — CHANGES_REQUESTED** [review]
You also need to update usages of `isText` to `isTextBased`.

**Comment by ImRodry** [comment]
CI will fail until #7215 is merged

**Review by kyranet — APPROVED** [review]

**Review by SpaceEEC — APPROVED** [review]

File: packages/discord.js/typings/index.d.ts
Comment (Jiralite) [inline packages/discord.js/typings/index.d.ts:1065]: I'm wondering if it's better to have this as the intersection `GuildBasedChannel & TextBasedChannel` instead? It makes it a lil' more dynamic then, right?

File: packages/discord.js/src/structures/Channel.js
Comment (Jiralite) [inline packages/discord.js/src/structures/Channel.js:129]:
```suggestion
   * Indicates whether this channel is a {@link PartialGroupDMChannel}.
```

File: packages/discord.js/src/structures/Channel.js
Comment (Jiralite) [inline packages/discord.js/src/structures/Channel.js:145]:
```suggestion
   * Indicates whether this channel is a {@link NewsChannel}.
```

File: packages/discord.js/src/structures/Channel.js
Comment (Jiralite) [inline packages/discord.js/src/structures/Channel.js:153]:
```suggestion
   * Indicates whether this channel is a {@link StoreChannel}.
```

**Review by Jiralite — COMMENTED** [review]
Some consistency suggestions and a TypeScript comment!

File: packages/discord.js/src/structures/Channel.js
Comment (ImRodry) [reply packages/discord.js/src/structures/Channel.js:129]: y tho

File: packages/discord.js/typings/index.d.ts
Comment (ImRodry) [reply packages/discord.js/typings/index.d.ts:1065]: Absolutely, will do

File: packages/discord.js/src/structures/Channel.js
Comment (Jiralite) [reply packages/discord.js/src/structures/Channel.js:129]: Because you've done it for some others, and that's just really... unsettling?

**Comment by iCrawl** [comment]
With #7077 merged you'll have to adapt your PR to it :+1:

File: packages/discord.js/src/structures/Channel.js
Comment (almeidx) [inline packages/discord.js/src/structures/Channel.js:6]:
```suggestion
const { ThreadChannelTypes } = require('../util/Constants');
```

File: packages/discord.js/src/structures/Channel.js
Comment (almeidx) [inline packages/discord.js/src/structures/Channel.js:166]:
```suggestion
    return ThreadChannelTypes.includes(this.type);
```

**Pull request merged**
